nab/detectors/bayes_changept/bayes_changept_detector.py

```diff
@@ -0,0 +1,164 @@
+"""
+Anomaly detection algorithm that implements Bayesian changepoint detection, as
+described in
+  Ryan P. Adams, David J.C. MacKay, Bayesian Online Changepoint Detection,
+arXiv 0710.3742 (2007).
+
+For the author's implementation (in MATLAB) please see
+http://hips.seas.harvard.edu/content/bayesian-online-changepoint-detection
+
+It has been modified from the author's code to run online: rather than
+initializing a run-length matrix with the size of the dataset, we use an array
+that recursively overwrites old data (no longer needed by the algorithm).
+"""
+from functools import partial
+import numpy
+from scipy import stats
+
+from nab.detectors.base import AnomalyDetector
+
+
+
+class BayesChangePtDetector(AnomalyDetector):
+
+  def __init__(self, *args, **kwargs):
+
+    super(BayesChangePtDetector, self).__init__(*args, **kwargs)
+
+    # Setup the matrix that will hold our beliefs about the current
+    # run lengths. We'll initialize it all to zero at first. For efficiency
+    # we preallocate a data structure to hold only the info we need to detect
+    # change points: columns for the current timestep t and t+1, and a
+    # sufficient number of rows (where each row represents probabilites of a
+    # run of that length).
+    self.maxRunLength = 500
+    self.runLengthProbs = numpy.zeros((self.maxRunLength + 2, 2))
+    # Time t=0 is a boundary condition, where we know the run length is 0.
+    self.runLengthProbs[0, 0] = 1.0
+
+    # Init variables for state.
+    self.timestep = 0
+    self.previousMaxRun = 1
+
+    # Define algorithm's helpers.
+    self.observationLikelihoood = StudentTDistribution(alpha=0.1,
+                                                       beta=0.001,
+                                                       kappa=1.0,
+                                                       mu=0.0)
+    self.hazardFunction = partial(constantHazard, lambdaConst=250)
```

Comment:
Why use a partial? I'd recommend avoiding closures as many people are unfamiliar with them. Highly recommend keeping this code simple to read. Just have self.lambdaConst as a parameter (just like the others) and just call constantHazard with self.lambdaConst.